### PR TITLE
Refactor producer goroutines somewhat

### PR DESCRIPTION
**Pull request by eapache**
- rename topicDispatcher to dispatcher
- make a topicProducer struct to handle the old partitionDispatcher
- make a partitionProducer struct to handle the old leaderDispatcher
- move and rename methods for a much more readable structure

Part 1 of #300.

There shouldn't be any functional changes. Thoughts @Shopify/kafka ?

### CODE REVIEW

File: async_producer.go [hunk 1]
```diff
@@ -200,16 +200,9 @@ func (p *asyncProducer) AsyncClose() {
 	go withRecover(p.shutdown)
 }
 
-///////////////////////////////////////////
-// In normal processing, a message flows through the following functions from top to bottom,
-// starting at topicDispatcher (which reads from Producer.input) and ending in flusher
-// (which sends the message to the broker). In cases where a message must be retried, it goes
-// through retryHandler before being returned to the top of the flow.
-///////////////////////////////////////////
-
```
Comment (wvanbergen) [inline async_producer.go]: We should probably replace this with something else.

Comment (eapache) [inline async_producer.go]: Ya. Once this is merged I'll try and find time for "part 2" which will be a similar process for the aggregator and flusher. At that point this comment can come back, excepti "flows through the following structs, via each `dispatch()` method".

File: async_producer.go [hunk 2]
```diff
@@ -264,133 +255,158 @@ func (p *asyncProducer) topicDispatcher() {
...
+	} else if choice < 0 || choice >= numPartitions {
+		return ErrInvalidPartition
 	}
 
+	msg.Partition = partitions[choice]
+
+	return nil
+}
+
+// one per partition per topic
```
Comment (wvanbergen) [inline async_producer.go]: Maybe we should say there's only one goroutine accessing the internals, so there's no need for concurrency guards. The only way to communicate with this struct is to use the `input` and `output` channels; all the other stuff should be considered "private". I guess there's no real way to enforce this, except by using an `internal` subpackage.

Comment (eapache) [inline async_producer.go]: ya, it's the same pattern we already used in the consumer... one goroutine per struct, communicate only via channels

not sure it needs explicit documentation in comments, the pattern is pretty obvious based on usage I think?

Comment (wvanbergen) [inline async_producer.go]: Might be nice to document this somewhere as a general design principle. Doesn't have to be in the code, wiki page would be fine too,

Comment (wvanbergen) [inline async_producer.go]: Actually, more precise would be: communicate between objects only using the constructor and method; never using struct members directly. (right?)

Comment (eapache) [inline async_producer.go]: I've played with a couple of different constructor patterns: here the caller passes in an input channel. For brokerConsumer, it's created as part of constructor then accessed as a member by the caller. For partitionConsumer, the constructor is embedded directly in the calling method...

I'm still not sure which I like best, but the overriding directive is: the struct has an associated goroutine, and only that goroutine may access non-channel members directly

Comment (wvanbergen) [inline async_producer.go]: I think I like the idea of creating a method for everything that should be accessible from outside the struct. e.g. by using a method that returns the internally created channel. That way struct members are always off limits for external objects.

Comment (eapache) [inline async_producer.go]: in all the cases so far, all the methods are off-limits too, they're internal-only helpers

**Comment by wvanbergen** [comment]
This looks fine to me.

Like we were discussing above, the main thing I am looking for is a way to communicate how these objects should be used. Whether that would be with docs, or somehow using a code style I don't know.

We can tackle this after step 2.

**Comment by eapache** [comment]
It's going to be part of my talk at Golang UK if nothing else :)

**Pull request merged**